### PR TITLE
Don't store contexts in lease pool structs

### PR DESCRIPTION
**Pull request by twz123**
## Description

The context shouldn't be stored in structs, but passed as the first argument to functions and methods. Remove the context storage from the lease pool and pass it to the methods where it's actually needed.

See:
* #3518

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## How Has This Been Tested?

- [ ] Manual test
- [ ] Auto test added

## Checklist:

- [x] My code follows the style [guidelines](https://github.com/k0sproject/k0s/blob/main/docs/contributors/overview.md) of this project 
- [x] My commit messages are [signed-off](https://github.com/k0sproject/k0s/blob/main/docs/contributors/github_workflow.md)
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by twz123** [comment]
This revealed and depends on #4744.

File: pkg/component/controller/controllersleasecounter.go [hunk 1]
```diff
@@ -69,18 +67,19 @@ func (l *K0sControllersLeaseCounter) Start(ctx context.Context) error {
...
-		leaderelection.WithLogger(log),
-		leaderelection.WithContext(ctx))
+	leasePool, err := leaderelection.NewLeasePool(client, leaseName, l.InvocationID,
+		leaderelection.WithLogger(log))
 	if err != nil {
 		return err
 	}
-	events, cancel, err := leasePool.Watch()
+
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (juanluisvaladas) [inline pkg/component/controller/controllersleasecounter.go:76]: I don't understand this change.
Why shouldn't the `context.WithCancel` be based on the one `Start`gets as an argument? If that context is Done I think this should be done too.

File: pkg/component/controller/leaderelector/leasepool.go [hunk 2]
```diff
@@ -61,19 +61,20 @@ func (l *LeasePool) Init(_ context.Context) error {
...
-	leasePool, err := leaderelection.NewLeasePool(ctx, client, l.name, l.invocationID,
-		leaderelection.WithLogger(l.log),
-		leaderelection.WithContext(ctx))
+	leasePool, err := leaderelection.NewLeasePool(client, l.name, l.invocationID,
+		leaderelection.WithLogger(l.log))
 	if err != nil {
 		return err
 	}
-	events, cancel, err := leasePool.Watch()
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (juanluisvaladas) [inline pkg/component/controller/leaderelector/leasepool.go:74]: Same thing here.

File: pkg/component/controller/controllersleasecounter.go (hunk 1 above)
Comment (twz123) [reply pkg/component/controller/controllersleasecounter.go:76]: This is one of the big issues that exists with the component model in k0s. See #4357 and its predecessor #1844.

**Review by juanluisvaladas — APPROVED** [review]

**Pull request merged**